# Introduction to Assembly Language

Let's learn some assembly language! We are going to learn how some instructions work. 

Here are the different types of instructions in assembly:
* Data Movement
* Arithmetic
* Control flow

Let's start with data movement instructions!

## Storing values into a register

**Registers** are the storage locations at the heart of the central processing unit. They can store constants, memory, and values from another register. 

The instruction *mov* handles data movement in various ways listed below.
* mov reg, constant
* mov reg, reg
* mov reg, [mem]
* mov [mem], constant
* mov [mem], reg

The value of the second operand is being stored into the first operand. 

In the cell below, the constant 2 is being stored in register EDX and the constant 16 is being stored in register EBX.

In [1]:
mov edx, 2
mov ebx, 16

Changes: Register Register Value Memory Location Memory Value EBX 16 EDX 2

## Arithmetic Instructions

In assembly, we have a set of arithmetic instructions that require one or two operands.

Instructions that require one operand:
* dec reg
    * Decrements the value in a register by 1 

* inc reg
    * Increments the value in a register by 1 
* neg reg
    * Negates the value in a register
* idiv reg
    * Divides the value in EDX:EAX by the specified operand value
    * Quotient is stored in EAX
    * Remainder is stored in EDX

Some intructions that required two operands:
* add reg, reg
* sub reg, constant
* imul reg, reg
The result of the arithmetic operation is overwrites the value of the first operand.

In the example below, EAX holds the value 10 and EBX holds the value 3. The product of EAX and EBX is stored into EAX. 

In [2]:
mov eax, 10
mov ebx, 3
imul eax, ebx

Changes: Register Register Value Memory Location Memory Value EAX 30 EBX 3 EDX 2

## Creating Functions 

To create a function in assembly, state a function name and write a sequence of assembly code. The function name represents as a locator to jump back to.

Below, the *sum* function is implemented. The *sum* function stores the sum of EAX and EBX in ECX.

In [3]:
mov eax, 10
mov ebx, 20

sum: mov ecx, eax
     add ecx, ebx

Changes: Register Register Value Memory Location Memory Value EAX 10 EBX 20 ECX 30 EDX 2

## Control Flow Instructions

In assembly language, we have control flow instructions that determine what the next instruction should be. 
Some of the instructions are determined by the value of the flags. 

### Comparing values
For example, the *cmp* instruction compares two values and sets the *zero* and *signed* flags.

In [ ]:
cmp ebx, 1

The *cmp* instruction compared the value of EBX, which was 20, and the constant 1. It first substracts
the constant from EBX. If the difference is zero, the *zero* flag is set to 1; otherwise, it is set to 0. 
Similarly, the *signed* flag is set to 1 if the difference is negative; otherwise, it is set to 0. 

Given that EBX - 1 is greater than zero, the *zero* and *signed* flags are set to 0.

In [ ]:
; An example of setting the zero flag to 1

cmp ebx, 20

### Jumping to Labels
Other control flow instructions involve jump instructions. Jump instructions tell assembly to jump to a function if specified requirements are met.

Some jump functions in assembly are: 
* jmp label
    * Jumps to a function 
* jg label
    * Known as: Jump if greater than
    * Jumps to a function if both *signed* and *zero* flags are 0.
* jne label
    * Jump if not equal to
    * Jumps to a function if the *zero* flag is 0.
    
Besides the *jmp* function, the jump instructions rely on the flag values, which implies that a *cmp* instruction should be executed prior to jumping.

Below a *multiply function* is implemented. It is performing 2 * 4, which is the same as 2 + 2 + 2 + 2.
Until EBX, which currently holds 4, is decremented to the value of 1, the value of EDX is incremented by the value of 2. 

In [ ]:
mov edx, 2
mov ecx, 2
mov ebx, 4

multiply: add edx, ecx
          dec ebx
          cmp ebx, 1
          jne multiply

### Calling and Returning from a Function

Similar to jumping to a label, we can also call a function. By a calling a label, we store the location in the
program where we did a *call*. 

The function that is called will have a *ret* instruction to indicate that we have reached the end of the 
function and will now jump back to where *call* was last executed. 

In the example program, we reuse the *multiply* function. Once EBX's value is 1, we no longer loop back to the beginning of the function. Instead we return back to where *call multiply* was executed. Then, we jump to the *end* function, where the result value is stored at memory location 0.

In [ ]:
mov edx, 2
mov ecx, 2
mov ebx, 4
call multiply
jmp end

multiply: add edx, ecx
          dec ebx
          cmp ebx, 1
          jne multiply
          ret
          
end: mov [0], edx

## Interrupt Functions

In assembly language, we also have interrupt functions! 

* int 22
    * We pretend that we are getting a keyboard input from the user.
* int 32
    * Exits the current program
    
For both interrupts to occur, EAX must hold a value of 0.

# Raising a Number to a Power

Now let's put everything together and write a program to raise a number to a power! 

In our example, we will raise 2 to the 16th power. Similar to our *multiply* function, raising 2<sup>4</sup> is the same as 2 * 2 * 2 * 2. 

We will have EBX hold our exponent value, which will decrement until it hits 1. EDX will hold our accumulative product after multiplying by 2 per function call.

A *power* function will be called to execute the looping and once the function is fully executed, it will return to the *call* execution. An interrupt will be called to exit the program.

In [4]:
; In edx, we put the number to raise to the power we put in ebx.
      mov edx, 2
      mov ebx, 16
      call power
      mov eax, 0
      int 32

power: mov ecx, edx
loop: imul edx, ecx
      dec ebx
      cmp ebx, 1
      jne loop
      ret

Changes: Register Register Value Memory Location Memory Value EAX 0 EBX 1 ECX 2 EDX 65536

You'll notice that register `EDX` contains the value 65536, which is indeed 2 raised to the 16th power.